File: app/src/main/java/com/youngfeng/android/assistant/server/webrtc/WebRTCTransport.kt
package com.youngfeng.android.assistant.server.webrtc

import android.content.Context
import android.content.Intent
import android.media.projection.MediaProjection
import com.youngfeng.android.assistant.server.transport.StreamTransport
import kotlinx.coroutines.*
import org.webrtc.*
import timber.log.Timber
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * WebRTC-based transport implementation
 * Provides high-performance, low-latency screen sharing with audio
 */
class WebRTCTransport(
    private val context: Context,
    private val mediaProjection: MediaProjection,
    private val signalingUrl: String
) : StreamTransport {

    private var peerConnectionFactory: PeerConnectionFactory? = null
    private var peerConnection: PeerConnection? = null
    private var videoTrack: VideoTrack? = null
    private var audioTrack: AudioTrack? = null
    private var videoCapturer: ScreenCapturerAndroid? = null

    private var signalingClient: SignalingClient? = null
    private val sessionId = java.util.UUID.randomUUID().toString()

    private var state: StreamTransport.ConnectionState = StreamTransport.ConnectionState.DISCONNECTED
    private var stateListener: ((StreamTransport.ConnectionState) -> Unit)? = null

    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // EGL context for hardware encoding
    private var eglBase: EglBase? = null

    // Screen dimensions
    private var screenWidth = 1080
    private var screenHeight = 1920

    companion object {
        private const val VIDEO_TRACK_ID = "video_track"
        private const val AUDIO_TRACK_ID = "audio_track"
        private const val STREAM_ID = "screen_stream"

        // Video encoding parameters
        private const val VIDEO_FPS = 60
        private const val VIDEO_RESOLUTION_WIDTH = 1080
        private const val VIDEO_RESOLUTION_HEIGHT = 1920
        private const val VIDEO_BITRATE_KBPS = 5000
    }

    override fun getType(): StreamTransport.Type = StreamTransport.Type.WEBRTC

    override suspend fun start() = withContext(Dispatchers.Main) {
        try {
            Timber.d("WebRTCTransport: Starting")
            updateState(StreamTransport.ConnectionState.CONNECTING)

            // Initialize WebRTC
            initializeWebRTC()

            // Create peer connection
            createPeerConnection()

            // Add media tracks
            addMediaTracks()

            // Connect to signaling server
            connectSignaling()

            Timber.d("WebRTCTransport: Started successfully")
        } catch (e: Exception) {
            Timber.e(e, "Failed to start WebRTCTransport")
            updateState(StreamTransport.ConnectionState.FAILED)
            throw e
        }
    }

    override fun stop() {
        Timber.d("WebRTCTransport: Stopping")

        // Disconnect signaling
        signalingClient?.disconnect()
        signalingClient = null

        // Stop video capturer
        try {
            videoCapturer?.stopCapture()
            videoCapturer?.dispose()
            videoCapturer = null
        } catch (e: Exception) {
            Timber.e(e, "Error stopping video capturer")
        }

        // Close peer connection
        peerConnection?.close()
        peerConnection = null

        // Dispose tracks
        videoTrack?.dispose()
        videoTrack = null
        audioTrack?.dispose()
        audioTrack = null

        // Dispose factory
        peerConnectionFactory?.dispose()
        peerConnectionFactory = null

        // Release EGL
        eglBase?.release()
        eglBase = null

        scope.cancel()

        updateState(StreamTransport.ConnectionState.DISCONNECTED)
        Timber.d("WebRTCTransport: Stopped")
    }

    override suspend fun sendVideoFrame(frameData: ByteArray) {
        // Not used in WebRTC - video is captured directly by VideoCapturer
    }

    override suspend fun sendAudioData(audioData: ByteArray) {
        // Not used in WebRTC - audio is captured directly by AudioTrack
    }

    override fun getConnectionState(): StreamTransport.ConnectionState = state

    override fun getConnectionCount(): Int {
        return if (peerConnection?.connectionState() == PeerConnection.PeerConnectionState.CONNECTED) 1 else 0
    }

    override fun hasConnections(): Boolean {
        return peerConnection?.connectionState() == PeerConnection.PeerConnectionState.CONNECTED
    }

    override fun setStateListener(listener: (StreamTransport.ConnectionState) -> Unit) {
        this.stateListener = listener
    }

    private fun initializeWebRTC() {
        Timber.d("WebRTCTransport: Initializing WebRTC")

        // Initialize PeerConnectionFactory
        val options = PeerConnectionFactory.InitializationOptions.builder(context)
            .setEnableInternalTracer(false)
            .createInitializationOptions()

        PeerConnectionFactory.initialize(options)

        // Create EGL context
        eglBase = EglBase.create()

        // Build PeerConnectionFactory
        val encoderFactory = DefaultVideoEncoderFactory(
            eglBase!!.eglBaseContext,
            true, // enableIntelVp8Encoder
            true // enableH264HighProfile
        )

        val decoderFactory = DefaultVideoDecoderFactory(eglBase!!.eglBaseContext)

        peerConnectionFactory = PeerConnectionFactory.builder()
            .setVideoEncoderFactory(encoderFactory)
            .setVideoDecoderFactory(decoderFactory)
            .setOptions(PeerConnectionFactory.Options())
            .createPeerConnectionFactory()

        Timber.d("WebRTCTransport: PeerConnectionFactory created")
    }

    private fun createPeerConnection() {
        Timber.d("WebRTCTransport: Creating PeerConnection")

        // ICE servers configuration
        val iceServers = listOf(
            PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer()
        )

        val rtcConfig = PeerConnection.RTCConfiguration(iceServers).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
            continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY
        }

        val observer = object : PeerConnection.Observer {
            override fun onIceCandidate(candidate: IceCandidate?) {
                candidate?.let {
                    Timber.d("WebRTCTransport: New ICE candidate")
                    signalingClient?.send(
                        SignalingMessage.IceCandidate(
                            sessionId = sessionId,
                            candidate = it.sdp,
                            sdpMid = it.sdpMid,
                            sdpMLineIndex = it.sdpMLineIndex
                        )
                    )
                }
            }

            override fun onConnectionChange(newState: PeerConnection.PeerConnectionState?) {
                Timber.d("WebRTCTransport: Connection state changed to $newState")
                when (newState) {
                    PeerConnection.PeerConnectionState.CONNECTED -> {
                        updateState(StreamTransport.ConnectionState.CONNECTED)
                    }
                    PeerConnection.PeerConnectionState.FAILED -> {
                        updateState(StreamTransport.ConnectionState.FAILED)
                    }
                    PeerConnection.PeerConnectionState.DISCONNECTED -> {
                        updateState(StreamTransport.ConnectionState.DISCONNECTED)
                    }
                    else -> {}
                }
            }

            override fun onIceConnectionChange(newState: PeerConnection.IceConnectionState?) {
                Timber.d("WebRTCTransport: ICE connection state: $newState")
            }

            override fun onIceConnectionReceivingChange(receiving: Boolean) {
                Timber.d("WebRTCTransport: ICE connection receiving: $receiving")
            }

            override fun onIceGatheringChange(newState: PeerConnection.IceGatheringState?) {
                Timber.d("WebRTCTransport: ICE gathering state: $newState")
            }

            override fun onAddStream(stream: MediaStream?) {
                Timber.d("WebRTCTransport: Stream added")
            }

            override fun onRemoveStream(stream: MediaStream?) {
                Timber.d("WebRTCTransport: Stream removed")
            }

            override fun onDataChannel(dataChannel: DataChannel?) {
                Timber.d("WebRTCTransport: Data channel received")
            }

            override fun onRenegotiationNeeded() {
                Timber.d("WebRTCTransport: Renegotiation needed")
            }

            override fun onSignalingChange(newState: PeerConnection.SignalingState?) {
                Timber.d("WebRTCTransport: Signaling state: $newState")
            }

            override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>?) {
                Timber.d("WebRTCTransport: ICE candidates removed")
            }

            override fun onAddTrack(receiver: RtpReceiver?, streams: Array<out MediaStream>?) {
                Timber.d("WebRTCTransport: Track added")
            }
        }

        peerConnection = peerConnectionFactory?.createPeerConnection(rtcConfig, observer)

        Timber.d("WebRTCTransport: PeerConnection created")
    }

    private fun addMediaTracks() {
        Timber.d("WebRTCTransport: Adding media tracks")

        // Create video track
        videoTrack = createVideoTrack()
        videoTrack?.let {
            peerConnection?.addTrack(it, listOf(STREAM_ID))
            Timber.d("WebRTCTransport: Video track added")
        }

        // Create audio track (if Android 10+)
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
            audioTrack = createAudioTrack()
            audioTrack?.let {
                peerConnection?.addTrack(it, listOf(STREAM_ID))
                Timber.d("WebRTCTransport: Audio track added")
            }
        } else {
            Timber.w("WebRTCTransport: Audio capture not supported on Android < 10")
        }
    }

    private fun createVideoTrack(): VideoTrack? {
        try {
            // Create screen capturer
            videoCapturer = ScreenCapturerAndroid(
                Intent(), // Will be set by MediaProjection
                object : MediaProjection.Callback() {}
            )

            // Create video source
            val surfaceTextureHelper = SurfaceTextureHelper.create(
                "CaptureThread",
                eglBase!!.eglBaseContext
            )

            val videoSource = peerConnectionFactory?.createVideoSource(false)
            videoCapturer?.initialize(surfaceTextureHelper, context, videoSource?.capturerObserver)

            // Start capture
            videoCapturer?.startCapture(VIDEO_RESOLUTION_WIDTH, VIDEO_RESOLUTION_HEIGHT, VIDEO_FPS)

            // Create video track
            return peerConnectionFactory?.createVideoTrack(VIDEO_TRACK_ID, videoSource)
        } catch (e: Exception) {
            Timber.e(e, "Failed to create video track")
            return null
        }
    }

    private fun createAudioTrack(): AudioTrack? {
        try {
            val audioConstraints = MediaConstraints()
            val audioSource = peerConnectionFactory?.createAudioSource(audioConstraints)
            return peerConnectionFactory?.createAudioTrack(AUDIO_TRACK_ID, audioSource)
        } catch (e: Exception) {
            Timber.e(e, "Failed to create audio track")
            return null
        }
    }

    private fun connectSignaling() {
        Timber.d("WebRTCTransport: Connecting to signaling server")

        val listener = object : SignalingClient.SignalingListener {
            override fun onConnected() {
                Timber.d("WebRTCTransport: Signaling connected")
                scope.launch {
                    createOffer()
                }
            }

            override fun onDisconnected() {
                Timber.d("WebRTCTransport: Signaling disconnected")
            }

            override fun onOffer(sdp: String) {
                Timber.d("WebRTCTransport: Received offer (unexpected for offerer)")
            }

            override fun onAnswer(sdp: String) {
                Timber.d("WebRTCTransport: Received answer")
                scope.launch {
                    handleAnswer(sdp)
                }
            }

            override fun onIceCandidate(candidate: String, sdpMid: String, sdpMLineIndex: Int) {
                Timber.d("WebRTCTransport: Received ICE candidate")
                scope.launch {
                    handleIceCandidate(candidate, sdpMid, sdpMLineIndex)
                }
            }

            override fun onError(error: String) {
                Timber.e("WebRTCTransport: Signaling error: $error")
                updateState(StreamTransport.ConnectionState.FAILED)
            }
        }

        signalingClient = SignalingClient(sessionId, signalingUrl, listener)
        signalingClient?.connect()
    }

    private suspend fun createOffer() = withContext(Dispatchers.Main) {
        try {
            Timber.d("WebRTCTransport: Creating offer")

            val constraints = MediaConstraints().apply {
                mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "false"))
                mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"))
            }

            val offer = suspendCancellableCoroutine<SessionDescription> { continuation ->
                peerConnection?.createOffer(
                    object : SdpObserver {
                        override fun onCreateSuccess(sdp: SessionDescription?) {
                            sdp?.let { continuation.resume(it, null) }
                                ?: continuation.resumeWithException(Exception("SDP is null"))
                        }

                        override fun onCreateFailure(error: String?) {
                            continuation.resumeWithException(Exception(error ?: "Unknown error"))
                        }

                        override fun onSetSuccess() {}
                        override fun onSetFailure(error: String?) {}
                    },
                    constraints
                )
            }

            // Set local description
            suspendCancellableCoroutine<Unit> { continuation ->
                peerConnection?.setLocalDescription(
                    object : SdpObserver {
                        override fun onSetSuccess() {
                            continuation.resume(Unit, null)
                        }

                        override fun onSetFailure(error: String?) {
                            continuation.resumeWithException(Exception(error ?: "Unknown error"))
                        }

                        override fun onCreateSuccess(sdp: SessionDescription?) {}
                        override fun onCreateFailure(error: String?) {}
                    },
                    offer
                )
            }

            // Send offer to signaling server
            signalingClient?.send(
                SignalingMessage.Offer(
                    sessionId = sessionId,
                    sdp = offer.description
                )
            )

            Timber.d("WebRTCTransport: Offer created and sent")
        } catch (e: Exception) {
            Timber.e(e, "Failed to create offer")
            updateState(StreamTransport.ConnectionState.FAILED)
        }
    }

    private suspend fun handleAnswer(sdp: String) = withContext(Dispatchers.Main) {
        try {
            Timber.d("WebRTCTransport: Setting remote description")

            val answer = SessionDescription(SessionDescription.Type.ANSWER, sdp)

            suspendCancellableCoroutine<Unit> { continuation ->
                peerConnection?.setRemoteDescription(
                    object : SdpObserver {
                        override fun onSetSuccess() {
                            Timber.d("WebRTCTransport: Remote description set successfully")
                            continuation.resume(Unit, null)
                        }

                        override fun onSetFailure(error: String?) {
                            Timber.e("WebRTCTransport: Failed to set remote description: $error")
                            continuation.resumeWithException(Exception(error ?: "Unknown error"))
                        }

                        override fun onCreateSuccess(sdp: SessionDescription?) {}
                        override fun onCreateFailure(error: String?) {}
                    },
                    answer
                )
            }
        } catch (e: Exception) {
            Timber.e(e, "Failed to handle answer")
        }
    }

    private suspend fun handleIceCandidate(candidate: String, sdpMid: String, sdpMLineIndex: Int) =
        withContext(Dispatchers.Main) {
            try {
                val iceCandidate = IceCandidate(sdpMid, sdpMLineIndex, candidate)
                peerConnection?.addIceCandidate(iceCandidate)
                Timber.d("WebRTCTransport: ICE candidate added")
            } catch (e: Exception) {
                Timber.e(e, "Failed to add ICE candidate")
            }
        }

    private fun updateState(newState: StreamTransport.ConnectionState) {
        state = newState
        stateListener?.invoke(newState)
    }
}

File: app/src/main/assets/web/js/webrtc-transport.js
/**
 * WebRTC-based transport for high-performance screen sharing
 */
class WebRTCTransport {
    constructor() {
        this.peerConnection = null;
        this.signalingWs = null;
        this.sessionId = null;
        this.videoElement = null;
        this.audioContext = null;
        this.isConnected = false;

        // Callbacks
        this.onVideoCallback = null;
        this.onAudioCallback = null;
        this.onConnectionStateChange = null;
    }

    async connect(signalingUrl, iceServers) {
        console.log('WebRTCTransport: Connecting to', signalingUrl);

        try {
            // Create peer connection
            this.peerConnection = new RTCPeerConnection({
                iceServers: iceServers || [
                    { urls: 'stun:stun.l.google.com:19302' }
                ]
            });

            // Setup peer connection handlers
            this.setupPeerConnectionHandlers();

            // Connect to signaling server
            await this.connectSignaling(signalingUrl);

            console.log('WebRTCTransport: Connected successfully');
        } catch (error) {
            console.error('WebRTCTransport: Connection failed:', error);
            throw error;
        }
    }

    setupPeerConnectionHandlers() {
        // Handle incoming media tracks
        this.peerConnection.ontrack = (event) => {
            console.log('WebRTCTransport: Received track:', event.track.kind);

            if (event.track.kind === 'video') {
                // Create MediaStream with video track
                const stream = new MediaStream([event.track]);

                if (this.onVideoCallback) {
                    this.onVideoCallback(stream);
                }
            } else if (event.track.kind === 'audio') {
                // Create MediaStream with audio track
                const stream = new MediaStream([event.track]);

                if (this.onAudioCallback) {
                    this.onAudioCallback(stream);
                }
            }
        };

        // Handle connection state changes
        this.peerConnection.onconnectionstatechange = () => {
            const state = this.peerConnection.connectionState;
            console.log('WebRTCTransport: Connection state:', state);

            this.isConnected = (state === 'connected');

            if (this.onConnectionStateChange) {
                this.onConnectionStateChange(state);
            }

            if (state === 'failed') {
                console.error('WebRTCTransport: Connection failed');
                this.disconnect();
            }
        };

        // Handle ICE connection state
        this.peerConnection.oniceconnectionstatechange = () => {
            console.log('WebRTCTransport: ICE connection state:',
                        this.peerConnection.iceConnectionState);
        };

        // Handle ICE gathering state
        this.peerConnection.onicegatheringstatechange = () => {
            console.log('WebRTCTransport: ICE gathering state:',
                        this.peerConnection.iceGatheringState);
        };

        // Handle ICE candidates
        this.peerConnection.onicecandidate = (event) => {
            if (event.candidate) {
                console.log('WebRTCTransport: New ICE candidate');
                this.sendSignalingMessage({
                    type: 'ice-candidate',
                    session_id: this.sessionId,
                    candidate: event.candidate.candidate,
                    sdp_mid: event.candidate.sdpMid,
                    sdp_m_line_index: event.candidate.sdpMLineIndex
                });
            }
        };
    }

    async connectSignaling(signalingUrl) {
        return new Promise((resolve, reject) => {
            this.signalingWs = new WebSocket(signalingUrl);

            this.signalingWs.onopen = () => {
                console.log('WebRTCTransport: Signaling connected');
                resolve();
            };

            this.signalingWs.onmessage = async (event) => {
                try {
                    const message = JSON.parse(event.data);
                    await this.handleSignalingMessage(message);
                } catch (error) {
                    console.error('WebRTCTransport: Error handling signaling message:', error);
                }
            };

            this.signalingWs.onerror = (error) => {
                console.error('WebRTCTransport: Signaling error:', error);
                reject(error);
            };

            this.signalingWs.onclose = () => {
                console.log('WebRTCTransport: Signaling disconnected');
            };
        });
    }

    async handleSignalingMessage(message) {
        console.log('WebRTCTransport: Received signaling message:', message.type);

        switch (message.type) {
            case 'ready':
                this.sessionId = message.session_id;
                console.log('WebRTCTransport: Session ID:', this.sessionId);
                break;

            case 'offer':
                await this.handleOffer(message.sdp);
                break;

            case 'ice-candidate':
                await this.handleIceCandidate(message);
                break;

            case 'error':
                console.error('WebRTCTransport: Signaling error:', message.error);
                break;

            default:
                console.warn('WebRTCTransport: Unknown message type:', message.type);
        }
    }

    async handleOffer(sdp) {
        try {
            console.log('WebRTCTransport: Handling offer');

            // Set remote description
            await this.peerConnection.setRemoteDescription(
                new RTCSessionDescription({ type: 'offer', sdp })
            );

            // Create answer
            const answer = await this.peerConnection.createAnswer();

            // Set local description
            await this.peerConnection.setLocalDescription(answer);

            // Send answer
            this.sendSignalingMessage({
                type: 'answer',
                session_id: this.sessionId,
                sdp: answer.sdp
            });

            console.log('WebRTCTransport: Answer sent');
        } catch (error) {
            console.error('WebRTCTransport: Error handling offer:', error);
        }
    }

    async handleIceCandidate(message) {
        try {
            const candidate = new RTCIceCandidate({
                candidate: message.candidate,
                sdpMid: message.sdp_mid,
                sdpMLineIndex: message.sdp_m_line_index
            });

            await this.peerConnection.addIceCandidate(candidate);
            console.log('WebRTCTransport: ICE candidate added');
        } catch (error) {
            console.error('WebRTCTransport: Error adding ICE candidate:', error);
        }
    }

    sendSignalingMessage(message) {
        if (this.signalingWs && this.signalingWs.readyState === WebSocket.OPEN) {
            this.signalingWs.send(JSON.stringify(message));
        } else {
            console.warn('WebRTCTransport: Signaling WebSocket not ready');
        }
    }

    disconnect() {
        console.log('WebRTCTransport: Disconnecting');

        // Send bye message
        if (this.signalingWs && this.signalingWs.readyState === WebSocket.OPEN) {
            this.sendSignalingMessage({
                type: 'bye',
                session_id: this.sessionId,
                reason: 'user_disconnect'
            });
        }

        // Close signaling
        if (this.signalingWs) {
            this.signalingWs.close();
            this.signalingWs = null;
        }

        // Close peer connection
        if (this.peerConnection) {
            this.peerConnection.close();
            this.peerConnection = null;
        }

        this.isConnected = false;
        console.log('WebRTCTransport: Disconnected');
    }

    renderVideo(videoElement) {
        this.videoElement = videoElement;

        this.onVideoCallback = (stream) => {
            videoElement.srcObject = stream;
            videoElement.play().catch(e => {
                console.error('WebRTCTransport: Error playing video:', e);
            });
        };

        this.onAudioCallback = (stream) => {
            // Audio is automatically played through the video element
            // or can be processed separately if needed
            console.log('WebRTCTransport: Audio stream received');
        };
    }

    async getStats() {
        if (!this.peerConnection) return null;

        const stats = await this.peerConnection.getStats();
        const result = {
            video: {},
            audio: {},
            connection: {}
        };

        stats.forEach(report => {
            if (report.type === 'inbound-rtp' && report.kind === 'video') {
                result.video = {
                    fps: report.framesPerSecond || 0,
                    bitrate: report.bytesReceived ?
                             (report.bytesReceived * 8 / report.timestamp) : 0,
                    packetsLost: report.packetsLost || 0,
                    jitter: report.jitter || 0
                };
            } else if (report.type === 'inbound-rtp' && report.kind === 'audio') {
                result.audio = {
                    bitrate: report.bytesReceived ?
                             (report.bytesReceived * 8 / report.timestamp) : 0,
                    packetsLost: report.packetsLost || 0
                };
            } else if (report.type === 'candidate-pair' && report.state === 'succeeded') {
                result.connection = {
                    currentRtt: report.currentRoundTripTime || 0,
                    availableOutgoingBitrate: report.availableOutgoingBitrate || 0
                };
            }
        });

        return result;
    }
}
